CryptoAnalysis/src/main/java/crypto/rules/CryptSLRuleReader.java

```diff
@@ -1,22 +1,154 @@
 package crypto.rules;
 
 import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
 import java.net.MalformedURLException;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Enumeration;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
 
 import crypto.cryptslhandler.CrySLModelReader;
 
 public class CryptSLRuleReader {
+
 	private static CrySLModelReader csmr;
-	
-	public static CryptSLRule readFromSourceFile(File file) {
-		if(csmr == null)
+
+	private static CrySLModelReader getReader(){
+		if (csmr == null)
+		{
 			try {
 				csmr = new CrySLModelReader();
-			} catch (MalformedURLException e) {
-				// TODO Auto-generated catch block
+			}
+			catch (MalformedURLException e){
 				e.printStackTrace();
+				// Sebastian:
+				// TODO: Current code could cause a NullPointerException
+				// Question: Is this Exception ever likely to happen?
+				// If no: Swallow it and 'accept' the NullPointerException
+				// If it can happen: Maybe we should re-throw a different exception (with this as the inner exception)
+				// Reason: In both cases (either NullPointerException or a custom) the creation of CryptSLRules
+				// is impossible if the MalformedURLException was thrown. So instead of allowing a generic NullPointerException
+				// we could throw something new that could be caught.
 			}
-		return csmr.readRule(file);
+		}
+		return csmr;
+	}
+
+	public static CryptSLRule readFromSourceFile(File file) {
+		return getReader().readRule(file);
 	}
-	
+
+    public static List<CryptSLRule> readFromDirectory(File directory) {
+        return readFromDirectory(directory, false);
+    }
+
+    public static List<CryptSLRule> readFromDirectory(File directory, boolean recursive) {
+        if (!directory.exists() || !directory.isDirectory())
+            return new ArrayList<>();
+
+        List<File> cryptSLFiles = new ArrayList<>();
+        findCryptSLFiles(directory, recursive, cryptSLFiles);
+        if (cryptSLFiles.size() == 0)
+            return new ArrayList<>();
+
+        CrySLModelReader reader = getReader();
+        List<CryptSLRule> rules = new ArrayList<>();
+        for (File file :cryptSLFiles)
+            rules.add(reader.readRule(file));
+
+        // TODO: Decide what happens with potential duplicates
+        return rules.stream().filter((x) -> x != null).collect(Collectors.toList());
+    }
+
+    // TODO: Discuss about .zip layout: Only Root or allow recursive.
+    public static List<CryptSLRule> readFromZipFile(File file) {
```

Comment:
Shouldn't this rather go into the other PR?

Reply:
it should and it is, i made some merging mistakes which i'm not interested to resolve. 
the plan is to first merge the other PR then this, so the changes you mention here will be removed from this PR anyway